### PR TITLE
C++11, Fix CI, fix use of deprecated timer, fix group assignment operator.

### LINKED ISSUE #91: Wdeprecated-copy warnings from group.hpp.
```
In file included from src/config/printer.cpp:19:
In file included from ./include/bitcoin/system/config/printer.hpp:25:
In file included from /home/runner/work/libbitcoin-system/libbitcoin-system/prefixenv/include/boost/format.hpp:48:
/home/runner/work/libbitcoin-system/libbitcoin-system/prefixenv/include/boost/format/group.hpp:59:12: warning: definition of implicit copy constructor for 'group1<boost::io::detail::group0>' is deprecated because it has a user-declared copy assignment operator [-Wdeprecated-copy]
   group1& operator=(const group1&);
           ^
/home/runner/work/libbitcoin-system/libbitcoin-system/prefixenv/include/boost/format/group.hpp:472:18: note: in implicit copy constructor for 'boost::io::detail::group1<boost::io::detail::group0>' first required here
group() { return detail::group1< detail::group0 > ( detail::group0() ); }
                 ^
  CXX      src/config/libbitcoin_system_la-script.lo
1 warning generated.
```
https://github.com/evoskuil/libbitcoin-system/runs/6664512291?check_suite_focus=true#step:6:2861

Origin: https://www.boost.org/doc/libs/1_79_0/boost/format/group.hpp

This warning repeats heavily in all of versions of boost, and there are no other warnings in our clang builds, which incorporate many boost libraries. There is no apparent workaround apart from global warning suppression.

### PR DESCRIPTION
**Pull request by jeking3**
Dependencies like optional and smart_ptr will no longer support C++03 therefore the new minimum language level is C++11, and all the CI jobs for both C++98 and C++03 have been removed.  Removed MSVC 2010 and 2012 from CI, moved some windows jobs to GHA.
    
Fixed use of the deprecated top level timer class as it was causing build problems using &lt;warnings-as-errors>.
    
Fixed -Wdeprecated-cast in group.hpp.
Fixed -Winvalid-source-encoding in sample_new_features.cpp.
Fixed -Wself-assign-overloaded in format_test2.cpp.
    
These changes also allow b2 to work properly when executed from within the format directory.

This closes #91.

### CODE REVIEW

**Comment by jeking3** [comment]
Please check the changes to group.hpp closely:

1. There was an issue building due to a declared but not implemented assignment operator.
2. It appeared to exist to make assignment impossible but it was only on group1 through group4.
3. I moved it into group0 and changed the structures to derive from the previous.
4. I tried using boost::noncopyable but that caused problems elsewhere, and my goal is to restore CI builds first.

File: Jamfile [hunk 1]
```diff
@@ -5,19 +5,12 @@
...
-      <warnings>all
-
-      <toolset>clang:<cxxflags>-Wextra
-      <toolset>clang:<cxxflags>-ansi
-      <toolset>clang:<cxxflags>-pedantic
-
-      <toolset>gcc:<cxxflags>-Wextra
-      <toolset>gcc:<cxxflags>-ansi
-      <toolset>gcc:<cxxflags>-pedantic
+        <warnings>on
```
Comment (pdimov) [inline Jamfile:12]: The previous warning level was essentially `<warnings>pedantic`; `<warnings>on` is much less strict.

File: include/boost/format/group.hpp [hunk 2]
```diff
@@ -34,9 +34,15 @@ namespace detail {
 
 
 // empty group, but useful even though.
-struct group0 
+struct group0
 {
-    group0()      {}
+    BOOST_FORCEINLINE group0() {}
+#if !defined(BOOST_NO_CXX11_DELETED_FUNCTIONS)
```
Comment (pdimov) [inline include/boost/format/group.hpp:40]: What compiler of the supported ones doesn't have deleted functions?

File: test/format_test2.cpp [hunk 3]
```diff
@@ -204,9 +204,13 @@ int main(int, char* [])
     BOOST_TEST_EQ((boost::format("%2.2s %-4.4s % 8.8s")
         % mystr % mystr % mystr).str(), "ab abcd  abcdefg");
 
-    // coverage, operator =
+    // coverage, operator =, copy constructor
     format fmt("%1%%2%%3%");
-    fmt = fmt;
+    format fmt2 = fmt;
+    format fmt3(fmt);
```
Comment (pdimov) [inline test/format_test2.cpp:210]: None of these test the assignment operator.

File: Jamfile (hunk 1 above)
Comment (jeking3) [reply Jamfile]: Thanks, I put pedantic back, and we'll see what complains in CI.

File: include/boost/format/group.hpp (hunk 2 above)
Comment (jeking3) [reply include/boost/format/group.hpp]: This was prior to setting the minimum language level to C++11 due to dependencies doing the same.

**Review by pdimov — APPROVED** [review]

**Pull request merged**